### PR TITLE
OSBB membership approval & OSBB neighbors list page

### PR DESCRIPTION
**Pull request by marikrebega**
[Trello Task: OSBB membership approval](https://trello.com/c/a2ihYpSt/73-osbb-membership-approval)
[Trello Task: OSBB neighbors list page](https://trello.com/c/jQl4HEus/88-osbb-neighbors-list-page)

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [ ] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [ ] "let!" is used in specs
- [x] screenshot of feature is added if task has UI

image
image
image

### CODE REVIEW

File: app/controllers/account/neighbors_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Account::NeighborsController < Account::AccountController
+  def index
+    @neighbors = User.where('osbb_id = ? and id != ?', current_user.osbb_id, current_user.id)
```
Comment (igorkasyanchuk) [inline app/controllers/account/neighbors_controller.rb:3]: а можна створити метод в юзері який вертає сусідів?

File: app/controllers/api/v1/neighbors_controller.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+class Api::V1::NeighborsController < Api::ApiController
+  include ControllerHelper
+  before_action :neighbor, only: %i[update]
+
+  def index
+    render json: User.where('osbb_id = ? and id != ?', current_user.osbb_id, current_user.id)
```
Comment (igorkasyanchuk) [inline app/controllers/api/v1/neighbors_controller.rb:6]: це що за копіпаст? створи той метод і заюзай його

File: app/controllers/account/neighbors_controller.rb (hunk 1 above)
Comment (igorkasyanchuk) [reply app/controllers/account/neighbors_controller.rb:3]: і де тест на цей метод?

File: app/javascript/components/neighbors/NeighborItem.jsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+          </div>
+        </div>
+      </div>
+    );
+  }
+}
+
+function Avatar(neighbor) {
+  if (neighbor.neighbor.avatar) {
+    return <img src={neighbor.neighbor.avatar.user_index.url} alt="Image" className="user-avatar" ></img>
```
Comment (chu4) [inline app/javascript/components/neighbors/NeighborItem.jsx:32]: neighbor.neighbor виглядає дивно

File: app/views/account/neighbors/index.json.jbuilder [hunk 4]
```diff
@@ -0,0 +1,4 @@
+json.neighbors @neighbors do |neighbor|
+  json.call(neighbor, :id, :full_name, :email, :approved)
+  json.avatar_url neighbor.avatar.user_index.url
```
Comment (chu4) [inline app/views/account/neighbors/index.json.jbuilder:3]: не розумію чого avatar.user_index.url ?

File: app/javascript/components/neighbors/NeighborItem.jsx (hunk 3 above)
Comment (marikrebega) [reply app/javascript/components/neighbors/NeighborItem.jsx:32]: refactored

File: app/views/account/neighbors/index.json.jbuilder (hunk 4 above)
Comment (marikrebega) [reply app/views/account/neighbors/index.json.jbuilder:3]: так дані приходять з моделі, то мабуть з carrierwave зв'язано
Знімок екрану з 2020-11-22 22-48-47

File: app/controllers/account/neighbors_controller.rb (hunk 1 above)
Comment (marikrebega) [reply app/controllers/account/neighbors_controller.rb]: додав

File: app/controllers/api/v1/neighbors_controller.rb (hunk 2 above)
Comment (marikrebega) [reply app/controllers/api/v1/neighbors_controller.rb]: виправив

File: app/models/user.rb [hunk 5]
```diff
@@ -17,6 +17,8 @@ class User < ApplicationRecord
   has_many :companies, through: :billing_contracts
   has_many :payments, through: :billing_contracts
   has_many :news, dependent: :nullify
+  has_many :neighbors, ->(user) { where('osbb_id = ? and users.id != ?', user.osbb_id, user.id) },
```
Comment (chu4) [inline app/models/user.rb:20]: no need for `osbb_id = ?` if you already have `through: :osbb `

Comment (chu4) [reply app/models/user.rb:20]: create `scope non_self` or smth like that 
to make this check `users.id != ?`

**Review by Ukhanskyi — APPROVED** [review]

File: app/models/user.rb (hunk 5 above)
Comment (marikrebega) [reply app/models/user.rb]: done

**Review by Andrii-Kozak — APPROVED** [review]

File: app/controllers/user/registrations_controller.rb [hunk 6]
```diff
@@ -30,6 +30,8 @@ def save_resource
       inactive_resource
     end
     send_mails
+    SignUpEmailSenderWorker.perform_async(resource.id)
+    User.last.reindex
```
Comment (chu4) [inline app/controllers/user/registrations_controller.rb:34]: why not resource.reindex?

File: app/controllers/api/v1/neighbors_controller.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    if @neighbor.update(neighbor_params)
+      @neighbor.reindex
+      render json: @neighbor
+    else
+      render json: @neighbor.errors, status: :unprocessable_entity
+    end
+  end
+
+  def search
+    @neighbors = User.search(params[:approved],
```
Comment (chu4) [inline app/controllers/api/v1/neighbors_controller.rb:19]: as approved is a boolean
would look better in where with osbb_id and id
search by some names phone number whould be better

File: app/controllers/user/registrations_controller.rb (hunk 6 above)
Comment (igorkasyanchuk) [reply app/controllers/user/registrations_controller.rb:34]: чому User.last
а якщо в цей самий час інший юзер зерагався, це вже буде інший юзер

Comment (marikrebega) [reply app/controllers/user/registrations_controller.rb:34]: refactor

File: app/controllers/api/v1/neighbors_controller.rb (hunk 7 above)
Comment (marikrebega) [reply app/controllers/api/v1/neighbors_controller.rb]: refactor

File: app/controllers/api/v1/neighbors_controller.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    else
+      render json: @neighbor.errors, status: :unprocessable_entity
+    end
+  end
+
+  def search
+    @neighbors = User.search(where: {
+                               approved: params[:approved],
+                               osbb_id: current_user.osbb_id,
+                               id: { not: current_user.id }
```
Comment (pereslop) [inline app/controllers/api/v1/neighbors_controller.rb:21]: you can use here scope .non_self

File: db/migrate/20201117125417_add_approved_to_users.rb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+class AddApprovedToUsers < ActiveRecord::Migration[6.0]
+  def up
+    add_column :users, :approved, :boolean
+    change_column_default :users, :approved, false
```
Comment (pereslop) [inline db/migrate/20201117125417_add_approved_to_users.rb:4]: `add_column :users, : approved, :boolean, :default => true` would not work? if yes, #change method should be enough

File: app/views/api/v1/neighbors/search.json.jbuilder [hunk 10]
```diff
@@ -0,0 +1,8 @@
+json.neighbors @neighbors do |neighbor|
```
Comment (pereslop) [inline app/views/api/v1/neighbors/search.json.jbuilder:1]: looks like this partial looks the same as above, cant you reuse only one?

**Review by pereslop — APPROVED** [review]

File: app/views/api/v1/neighbors/search.json.jbuilder (hunk 10 above)
Comment (marikrebega) [reply app/views/api/v1/neighbors/search.json.jbuilder:1]: refactored

File: db/migrate/20201117125417_add_approved_to_users.rb (hunk 9 above)
Comment (marikrebega) [reply db/migrate/20201117125417_add_approved_to_users.rb:4]: strong_migration gem does not miss such migration

**Review by marikrebega — COMMENTED** [review]
ok

**Review by chu4 — APPROVED** [review]

**Review by mmarusyk — APPROVED** [review]

**Pull request merged**